_posts/2021-01-11-pisanie-pluginow-do-intellij.md

```diff
@@ -0,0 +1,350 @@
+                    || Arrays.stream(psiClass.getFields()).anyMatch(field -> field.hasAnnotation("com.consdata.doc.Doc"))
+                    || Arrays.stream(psiClass.getMethods()).anyMatch(field -> field.hasAnnotation("com.consdata.doc.Doc"));
+        }
+        return false;
+    }
+
+    private String getDoc(PsiClass psiClass) {
+        Map<String, String> doc = new HashMap<>();
+        doc.put(psiClass.getName(), evaulate(psiClass.getProject(), psiClass.getAnnotation("com.consdata.doc.Doc").findAttributeValue("value")));
+
+        doc.putAll(Arrays.stream(psiClass.getFields())
+                .filter(psiField -> psiField.hasAnnotation("com.consdata.doc.Doc"))
+                .collect(Collectors.toMap(PsiField::getName, psiField -> evaulate(psiField.getProject(), psiField.getAnnotation("com.consdata.doc.Doc").findAttributeValue("value")))));
+
+        doc.putAll(Arrays.stream(psiClass.getMethods())
+                .filter(psiMethod -> psiMethod.hasAnnotation("com.consdata.doc.Doc"))
+                .collect(Collectors.toMap(PsiMethod::getName, psiMethod -> evaulate(psiMethod.getProject(), psiMethod.getAnnotation("com.consdata.doc.Doc").findAttributeValue("value")))));
+
+        return doc.entrySet().stream().map(e -> "<b>" + e.getKey() + "</b>: " + e.getValue()).collect(Collectors.joining("<br/>"));
+
+    }
+
+    private String evaulate(Project project, PsiAnnotationMemberValue expression) {
+        return JavaPsiFacade.getInstance(project).getConstantEvaluationHelper().computeConstantExpression(expression).toString();
+    }
+    
+    @Nullable
+    @Override
+    public List<String> getUrlFor(PsiElement element, PsiElement originalElement) {
+        return Collections.singletonList("");
+    }
+
+    @Override
+    public boolean canPromptToConfigureDocumentation(PsiElement psiElement) {
+        return false;
+    }
+
+    @Override
+    public void promptToConfigureDocumentation(PsiElement psiElement) {
+    }
+}
+```
+W `fetchExternalDocumentation` generujemy dokumentację dla wskazanej klasy analogicznie jak w akcji, tylko dla jednej klasy. Dodatkowo wzbogacamy dokumentację o znacznik HTML(w oknie `QuickDocumentation` rednerowany jest HTML). W `hasDocumentationFor` decydujemy czy dla danego elementu powinniśmy wygenerować dodatkową dokumentację, podobnie jak w przypadku grupy, gdy decydowaliśmy o jej aktywności.
+Tak zdefiniowany provider możemy teraz zdefiniować w manifeście pluginu jak rozszerzenie dla `documentationProvider`:
+
+
+```xml
+    <extensions defaultExtensionNs="com.intellij">
+        <documentationProvider implementation="com.consdata.article.DocProvider"/>
+    </extensions>
+```
+
+Po uruchomieniu pluginu, przechodzimy do klasy `UseCase`. Najeżdżamy na typ `TestClass` i wybieramy Ctrl + Q. Po wywołaniu skrótu pojawi się popup z wygnerowaną dokumentacją:
+
+![message](/assets/img/posts/2021-01-11-pisanie-pluginow-do-intellij/quickPopup.png)
+
+Po ponownym użyciu skrótu odtworzy się szuflada:
```

Comment:
tutaj się upewnię, bo nie wiem jaki miałeś zamysł:
szuflada ma się faktycznie "odtworzyć" czy miałeś na myśli "otworzyć". Pytam, bo gdy rozumuję to, tak jak ja się domyślam, to do szuflady pasuje mi bardziej otwieranie.